apache/cloudstack
# Persistent Networks Resources created on host during time of connecti…

…on, and cleanup during time of removal

### Description



This PR is a GSoC project: https://github.com/apache/cloudstack/issues/4814

This PR adds a feature to L2 persitent networks, where resources of the persistent networks are created at hostConnect phase of DefaultHostListener, and cleaned up at hostAboutToBeRemoved phase. 



This PR also fixes a bug, where rebooting the hosts removes the persistent network's resources, without readding them when the host is readded to the zone.

Fixes: #5196 









### Types of changes


### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale


#### Bug Severity



### Screenshots (if appropriate):


### How Has This Been Tested?
This has been manually tested with a mbx KVM, XenServer and VmWare setup on a local machine. 

Different manual test scenarios:
1. When the persistent network is created and the host is added afterwards, the newly added host has the persistent network's resources.
2. When a host within a zone with a persistent network is removed from the zone, its resources will be cleaned up.
3. When a host with a running VM is removed from the zone, the running VM will migrate before the resources are cleaned up.
4. When a host is rebooted, it still has access to the persistent network's resources.

A Marvin component test under 
test/integration/component/test_persistent_networks.py
added that tests the following steps for a KVM environment.
1. identify hosts in the zone, and remove the first
2. create a L2 persistent network
3. add the host back to the zone

Validation steps:
1. Persistent network state should be implemented before adding the host
2. Host should be added back in successfully
3. Host should have the persistent networks resources after being added




Test output results should be:

==== Marvin Init Started ====

=== Marvin Parse Config Successful ===

=== Marvin Setting TestData Successful===

==== Log Folder Path: /tmp/MarvinLogs/<LogId>. All logs will be available here ====

=== Marvin Init Logging Successful===

==== Marvin Init Successful ====
====Trying SSH Connection: Host:172.20.0.121 User:root                                   Port:22 RetryCnt:10===
===SSH to Host 172.20.0.121 port : 22 SUCCESSFUL===
{Cmd: ip addr | grep breth0-991 > /dev/null 2>&1; echo $? via Host: 172.20.0.121} {returns: ['0']}
=== TestName: test_newly_added_host_for_persistent_network_resources | Status : SUCCESS ===

===final results are now copied to: /tmp//MarvinLogs/test_persistent_networks_<LogId>===
